### PR TITLE
lightdm-mobile-greeter: init at 2022-10-30

### PR DESCRIPTION
**Pull request by uninsane**
###### Description of changes

this PR adds the `lightdm-mobile-greeter` package and then defines a module to simplify enabling of it. built on x86 and aarch64; tested on my aarch64 Pinephone.

lightdm-mobile-greeter is a greeter designed for mobile hardware. although most mobile DEs are capable of running without an external greeter by using their own lock screens, those tend to have different limitations. for example, the lock screen in Phosh doesn't do PAM session management -- but by using this lightdm-mobile-greeter which integrates with PAM, one can do things like decrypt a home directory or unlock keys on login.

the upstream project lacks many images/videos, so here's what the software looks like: <https://www.youtube.com/watch?v=whcFag0drLk>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bobby285271** [comment]
Are there any plans for upstream to bump their libhandy version? `libhandy_0` is something I really want to remove [as soon as fractal made a new version](https://github.com/NixOS/nixpkgs/pull/148088#issuecomment-1151062578)

**Comment by uninsane** [comment]
> Are there any plans for upstream to bump their libhandy version? Actually `libhandy_0` is something I want to remove [as soon as fractal made a new version](https://github.com/NixOS/nixpkgs/pull/148088#issuecomment-1151062578) 🙈

i don’t think the original developer (Raatty) has plans for this. but during the course of packaging for Nix i had to update other things in Cargo.lock and Raatty was quick to respond when i sent them a small patch.

so if the rest of the package looks good to everyone, i could take a stab at removing the libhandy_0 dep later this week & forward that upstream. i have no sense of how involved that is, but the project as a whole isn’t too large, so hopefully it’s manageable. would you be happy with the PR otherwise if i were to do this?

File: pkgs/applications/display-managers/lightdm-mobile-greeter/default.nix [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    mkdir -p $out/share/applications
+    substitute lightdm-mobile-greeter.desktop \
+      $out/share/applications/lightdm-mobile-greeter.desktop \
+      --replace lightdm-mobile-greeter $out/bin/lightdm-mobile-greeter
+  '';
+
+  passthru.xgreeters = linkFarm "lightdm-mobile-greeter-xgreeters" [{
+    path = "${lightdm-mobile-greeter}/share/applications/lightdm-mobile-greeter.desktop";
+    name = "lightdm-mobile-greeter.desktop";
+  }];
```
Comment (bobby285271) [inline pkgs/applications/display-managers/lightdm-mobile-greeter/default.nix:44]: It is unclear to me why we need to move the desktop file in `$out/share/applications` (for a lightdm greeter).

Can we just install the desktop files in `$out/share/xgreeters`, `substituteInPlace` that file and use it in `passthru.xgreeters`?

**Review by bobby285271 — COMMENTED** [review]
> i could take a stab at removing the libhandy_0 dep later this week & forward that upstream

Thanks for doing that!

Did not test, but the diff mostly LGTM so far, some nits:

**Comment by uninsane** [comment]
feedback applied and deployed/tested.

on the libhandy front: the rust library lightdm-mobile-greeter uses to bind libhandy supports libhandy 1.0+, so the work to update that should be doable without me touching any code outside the lightdm-mobile-greeter repo (good).

i tried swapping libhandy_0 for libhandy on the nix side and bumping the corresponding Cargo.toml dep: bindings are generated fine, but because of API differences the application needs some code changes in order to compile. figure it’ll take me a week or two to sort that out.

i don’t expect there’s much urgency to getting this PR mainlined into nixpkgs, so do as you see fit here. if this PR is still open when i get the new libhandy working i’ll amend it, if it’s been merged i’ll just open a new PR for the update.

**Pull request merged**